replicator/run.py

```diff
@@ -0,0 +1,161 @@
+        #  On the other hand, if several changes are done quickly on the same domain (e.g. delete/recreate domain),
+        #  fetching the new catalog here allows skipping useless intermediate states. Also, in case processing somehow
+        #  is stuck because the catalog is invalid, always fetching a new catalog improves chances of recovery.
+        #  Also, if this is not always done, then long-running AXFRs may hold up new changes to the catalog, as we
+        #  would be waiting for large zone's serial to update until we retrieve a new catalog.
+        catalog = pdns_request('get', path=f'/zones/{catalog_domain_id}').json()
+        local_catalog_serial = catalog['serial']
+
+        # Update the catalog in the background if necessary
+        remote_catalog_serial = query_serial(catalog_domain)
+        if local_catalog_serial != remote_catalog_serial:
+            print(f'Queueing catalog AXFR from {local_catalog_serial} to {remote_catalog_serial} ...')
+            pdns_request('put', path=f'/zones/{catalog_domain_id}/axfr-retrieve')
+
+        # See if the last parsed catalog is the current one
+        if serials[catalog_domain] == local_catalog_serial:
+            print(f'Nothing to do (catalog {serials[catalog_domain]} with {len(serials)} zones up to date).')
+            sleep(1)
+            continue
+
+        # Parse current state of catalog zone and determine slave-side changes
+        catalog_map = {}
+        for rrset in catalog['rrsets']:
+            m = re.match(r'^(serial)?\.?([0-9a-f]{40}).zones.catalog.internal.$', rrset['name'])
+            if m is None:
+                continue
+
+            subname, m_unique = m.groups()
+
+            map = catalog_map.setdefault(m_unique, {})
+            if rrset['type'] == 'PTR' and subname is None:
+                map['name'] = rrset['records'][0]['content']
+            elif rrset['type'] == 'TXT' and subname == 'serial':
+                map['serial'] = int(rrset['records'][0]['content'][1:-1])
+            else:
+                raise Exception('Unexpected catalog rrset %s' % rrset)
+
+        if not all(len(map) == 2 for map in catalog_map.values()):
+            raise Exception('Catalog has a value of m_unique with inconsistent content')
+
+        local_zones = set(name for name in serials.keys() if name != catalog_domain)
+        catalog_zones = set(map['name'] for map in catalog_map.values())
+
+        deletions = local_zones - catalog_zones
+        additions = catalog_zones - local_zones
+        modifications = {map['name']: map['serial'] for map in catalog_map.values()
+                         if map['name'] in serials and map['serial'] != serials[map['name']]}
+
+        for zone, serial in modifications.items():
+            # Refresh local serial in case an earlier AXFR has finished by now
+            serials[zone] = pdns_request('get', path='/zones/{}'.format(pdns_id(zone))).json()['serial']
```

Comment:
This can result in an infinite loop if the AXFR takes longer than our loop needs to return to the same point, unless pdns takes care of such duplicates.

Reply:
it does, extra AXFR requests are ignored (verified by Habbie in IRC)